File: 卡码/代码随想录/Go/二叉树/04-二叉搜索树属性/LeetCode_0501.go
//501. 二叉搜索树中的众数
package main

import (
	. "binaryTree/types"
)

func findMode(root *TreeNode) []int {
	var pre *TreeNode=nil
	var res []int=make([]int,0)
	var count,maxCount=0,0
	var traversal func(root *TreeNode)
	traversal=func(root *TreeNode){
		if root==nil {
			return
		}
		traversal(root.Left)
		if pre==nil||pre.Val!=root.Val {
			count=1
		}else{
			count++
		}
		pre=root
		if count==maxCount {
			res=append(res,root.Val)
		}else if count>maxCount {
			maxCount=count
			res=[]int{root.Val}
		}
		traversal(root.Right)
		return
	}
	traversal(root)
	return res
}

File: 卡码/代码随想录/1/06-二叉树/Java/binaryTree/src/main/java/com/kurumi/binarytree/_04_二叉搜索树属性/LeetCode_0501.java
//501. 二叉搜索树中的众数
package com.kurumi.binarytree._04_二叉搜索树属性;

import com.kurumi.binarytree.TreeNode;
import java.util.ArrayList;
import java.util.List;

public class LeetCode_0501 {

}
class Solution {
  List<Integer> res=new ArrayList<>();
  TreeNode pre=null;
  int maxCount=0;
  int count=0;
  public void traversal(TreeNode root){
    if(root==null) return;
    traversal(root.left);
    if(pre==null) count=1;
    else if(root.val==pre.val) count++;
    else count=1;
    pre=root;
    if(count==maxCount) res.add(root.val);
    if(count>maxCount){
      maxCount=count;
      res.clear();
      res.add(root.val);
    }
    traversal(root.right);
    return;
  }
  public int[] findMode(TreeNode root) {
    traversal(root);
    int[] result=new int[res.size()];
    for(int i=0;i<result.length;i++){
      result[i]=res.get(i);
    }
    return result;
  }
}
